Source/NSPathControl.m

```diff
@@ -0,0 +1,396 @@
+   
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+   
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; if not, write to the Free
+   Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110 USA.
+*/
+
+#import "AppKit/NSPathControl.h"
+#import "AppKit/NSPathCell.h"
+#import "AppKit/NSGraphics.h"
+#import "AppKit/NSDragging.h"
+#import "AppKit/NSPasteboard.h"
+#import "AppKit/NSMenu.h"
+#import "AppKit/NSOpenPanel.h"
+
+@implementation NSPathControl
+
++ (void) initialize
+{
+  if (self == [NSPathControl class])
+    {
+      [self setVersion: 1.0];
+      [self setCellClass: [NSPathCell class]];
+    }
+}
+
+- (instancetype) init
+{
+  self = [super init];
+  if (self != nil)
+    {
+      [self setPathStyle: NSPathStyleStandard];
+      [self setPathComponentCells: nil];
+      [self setURL: nil];
+      [self setDelegate: nil];
+      [self setAllowedTypes: [NSArray arrayWithObject: NSFilenamesPboardType]];
+    }
+  return self;
+}
+
+- (void) setPathStyle: (NSPathStyle)style
+{
+  [_cell setPathStyle: style];
+  [self setNeedsDisplay];
+}
+
+- (NSPathStyle) pathStyle
+{
+  return [_cell pathStyle];
+}
+
+- (NSPathComponentCell *) clickedPathComponentCell
+{
+  return [_cell clickedPathComponentCell];
+}
+
+- (NSArray *) pathComponentCells
+{
+  return [_cell pathComponentCells];
+}
+
+- (void) setPathComponentCells: (NSArray *)cells
+{
+  [_cell setPathComponentCells: cells];
+  [self setNeedsDisplay];
+}
+
+- (SEL) doubleAction;
+{
+  return [_cell doubleAction];
+}
+
+- (void) setDoubleAction: (SEL)doubleAction
+{
+  [_cell setDoubleAction: doubleAction];
+}
+
+- (NSURL *) URL
+{
+  return [_cell URL];
+}
+
+- (void) setURL: (NSURL *)url
+{
+  [_cell setURL: url];
+  [self setNeedsDisplay];
+}
+
+- (id<NSPathControlDelegate>) delegate
+{
+  return _delegate;
+}
+
+- (void) setDelegate: (id<NSPathControlDelegate>) delegate
+{
+  _delegate = delegate;
+}
+
+- (NSDragOperation) draggingSourceOperationMaskForLocal: (BOOL)flag
+{
+  if (flag)
+    {
+      return _localMask;
+    }
+
+  return _remoteMask;
+}
+
+- (void) setDraggingSourceOperationMask: (NSDragOperation)mask 
+                               forLocal: (BOOL)local
+{
+  if (local)
+    {
+      _localMask = mask;
+    }
+  else
+    {
+      _remoteMask = mask;
+    }
+}
+
+- (NSMenu *) menu
+{
+  return [super menu];
+}
+
+- (void) setMenu: (NSMenu *)menu
+{
+  [super setMenu: menu];
+}
+
+- (NSArray *) allowedTypes;
+{
+  return [_cell allowedTypes];
+}
+
+- (void) setAllowedTypes: (NSArray *)allowedTypes
+{
+  [_cell setAllowedTypes: allowedTypes];
+  [self registerForDraggedTypes: allowedTypes];
+}
+
+- (NSPathControlItem *) clickedPathItem
+{
+  return nil;
+}
+
+- (NSArray *) pathItems
```

Comment:
Shouldn't this come from the cell?

Reply:
Does now.

Reply:
I thought that you had removed these methods from the cell, as Apple does not declare them there.
I am fine with both ways, we just should not keep the values twice.